coreneuron/io/nrn_checkpoint.cpp

```diff
@@ -585,30 +510,118 @@ static void write_phase2(NrnThread& nt, FileHandlerWrap& fh) {
-double restore_time(const char* restore_dir) {
-    double rtime = 0;
-    if (strlen(restore_dir)) {
-        std::ostringstream filename;
-        FileHandler f;
-        filename << restore_dir << "/time.dat";
-        f.open(filename.str(), std::ios::in);
-        f.read_array(&rtime, 1);
-        f.close();
+// A call to finitialize must be avoided after restoring the checkpoint
+// as that would change all states to a voltage clamp initialization.
+// Nevertheless t and some spike exchange and other computer state needs to
+// be initialized.
+// Also it is occasionally the case that nrn_init allocates data so we
+// need to call it but avoid the internal call to initmodel.
+// Consult finitialize.c to help decide what should be here
+bool CheckPoints::initialize() {
+    dt2thread(-1.);
+    nrn_thread_table_check();
+    nrn_spike_exchange_init();
+
+    // in case some nrn_init allocate data we need to do that but do not
+    // want to call initmodel.
+    _nrn_skip_initmodel = true;
+    for (int i = 0; i < nrn_nthread; ++i) {  // should be parallel
+        NrnThread& nt = nrn_threads[i];
+        for (NrnThreadMembList* tml = nt.tml; tml; tml = tml->next) {
+            Memb_list* ml = tml->ml;
+            mod_f_t s = corenrn.get_memb_func(tml->index).initialize;
+            if (s) {
+                (*s)(&nt, ml, tml->index);
+            }
+        }
     }
-    return rtime;
+    _nrn_skip_initmodel = false;
+
+    // if PatternStim exists, needs initialization
+    for (NrnThreadMembList* tml = nrn_threads[0].tml; tml; tml = tml->next) {
+        if (tml->index == patstimtype && patstim_index >= 0 && patstim_te > 0.0) {
+            Memb_list* ml = tml->ml;
+            checkpoint_restore_patternstim(patstim_index,
+                                           patstim_te,
+                                           /* below correct only for AoS */
+                                           0,
+                                           ml->nodecount,
+                                           ml->data,
+                                           ml->pdata,
+                                           ml->_thread,
+                                           nrn_threads,
+                                           0.0);
+            break;
+        }
+    }
+
+    // Check that bbcore_write is defined if we want to use checkpoint
+    for (NrnThreadMembList* tml = nrn_threads[0].tml; tml; tml = tml->next) {
+        auto type = tml->index;
+        if (corenrn.get_bbcore_read()[type] && !corenrn.get_bbcore_write()[type]) {
+            auto memb_func = corenrn.get_memb_func(type);
+            fprintf(stderr,
+                    "Checkpoint is requested involving BBCOREPOINTER but there is no bbcore_write"
+                    " function for %s\n",
+                    memb_func.sym);
+            assert(corenrn.get_bbcore_write()[type]);
+        }
+    }
+
+
+    return restored;
 }
```

Comment:
If I am not mistaken, most of the changes in this `nrn_checkpoint.cpp` is just code rearrangement and there is no real logic change? (just to be sure!)